### PR TITLE
feat: support changing IndexPart::metadata_bytes to json in future release

### PR DESCRIPTION
**Pull request by koivunej**
## Problem

Currently we serialize the `TimelineMetadata` into bytes to put it into `index_part.json`. This `Vec<u8>` (hopefully `[u8; 512]`) representation was chosen because of problems serializing TimelineId and Lsn between different serializers (bincode, json). After #5335, the serialization of those types became serialization format aware or format agnostic.

We've removed the pageserver local `metadata` file writing in #6769.

## Summary of changes

Allow switching from the current serialization format to plain JSON for the legacy TimelineMetadata format in the future by adding a competitive serialization method to the current one (`crate::tenant::metadata::modern_serde`), which accepts both old bytes and new plain JSON.

The benefits of this are that dumping the index_part.json with pretty printing no longer produces more than 500 lines of output, but after enabling it produces lines only proportional to the layer count, like:

```json
{
  "version": ???,
  "layer_metadata": { ... },
  "disk_consistent_lsn": "0/15FD5D8",
  "legacy_metadata": {
    "disk_consistent_lsn": "0/15FD5D8",
    "prev_record_lsn": "0/15FD5A0",
    "ancestor_timeline": null,
    "ancestor_lsn": "0/0",
    "latest_gc_cutoff_lsn": "0/149FD18",
    "initdb_lsn": "0/149FD18",
    "pg_version": 15
  }
}
```

This is an alternative to #7663, which still uses the `Vec<u8>` idea, but this time with mixed bincode+json instead of just bincode.

In the future, I propose we completely stop using this legacy metadata type and wasting time trying to come up with another version numbering scheme in addition to the informative-only one already found in `index_part.json`, and go ahead with storing metadata or feature flags on the `index_part.json` itself.

#7699 is the "one release after" changes which starts to produce metadata in the index_part.json as json.

### CODE REVIEW

File: pageserver/src/tenant/metadata.rs [hunk 1]
```diff
@@ -276,13 +276,163 @@ impl Serialize for TimelineMetadata {
...
+        struct Visitor;
+
+        impl<'d> serde::de::Visitor<'d> for Visitor {
+            type Value = TimelineMetadata;
+
+            fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                f.write_str("BeSer bytes or json structure")
+            }
+
+            fn visit_seq<A>(self, seq: A) -> Result<Self::Value, A::Error>
```
Comment (skyzh) [inline pageserver/src/tenant/metadata.rs:307]: This deserialization method indicates that we cannot have array/map inside the timeline metadata? (It's fine but a thing to keep in mind).

**Review by skyzh — APPROVED** [review]

File: pageserver/src/tenant/metadata.rs (hunk 1 above)
Comment (skyzh) [reply pageserver/src/tenant/metadata.rs:307]: i thought this will affect nested fields, but it actually doesn't; mark as resolved

**Comment by koivunej** [comment]
We had some discussion perhaps last week if we want to keep the checksum or not in the json. The next PR will include refactoring which makes it trivial to include but, I'd rather not reorder the commits.

File: pageserver/src/tenant/metadata.rs [hunk 2]
```diff
@@ -276,13 +276,163 @@ impl Serialize for TimelineMetadata {
...
+        fn flush(&mut self) -> std::io::Result<()> {
+            Ok(())
+        }
+    }
+
+    #[derive(thiserror::Error)]
+    #[error("re-serializing for crc32 failed")]
+    struct Crc32CalculationFailed<E>(#[source] E);
+
+    // this should be false for a week, after that we can change it to true
```
Comment (koivunej) [inline pageserver/src/tenant/metadata.rs:370]:
```suggestion
    // this should be true for one release, after that we can change it to false
```

**Pull request merged**